File: v2/src/main/mongo/explain.ts
import { EJSON } from 'bson';
import type { MongoClient } from 'mongodb';
import type { ExplainRequest } from '../../shared/schema.js';
import { parseFilter, parseProjection, parseSort } from './queryParser.js';

export async function explainFind(
  client: MongoClient,
  req: ExplainRequest,
): Promise<string> {
  const filter = parseFilter(req.filter);
  const projection = parseProjection(req.projection);
  const sort = parseSort(req.sort);

  let cursor = client.db(req.db).collection(req.collection).find(filter);
  if (Object.keys(projection).length > 0) cursor = cursor.project(projection);
  if (Object.keys(sort).length > 0) cursor = cursor.sort(sort);
  if ((req.skip ?? 0) > 0) cursor = cursor.skip(req.skip!);
  if ((req.limit ?? 0) > 0) cursor = cursor.limit(req.limit!);

  const explain = await cursor.explain(req.verbosity ?? 'queryPlanner');
  return EJSON.stringify(explain, { relaxed: false });
}

File: v3/src/main/kotlin/io/mex/mongo/Explain.kt
package io.mex.mongo

import com.mongodb.client.MongoClient
import com.mongodb.ExplainVerbosity
import org.bson.json.JsonMode
import org.bson.json.JsonWriterSettings

private val EJSON: JsonWriterSettings = JsonWriterSettings.builder()
    .outputMode(JsonMode.EXTENDED)
    .build()

fun explainFind(client: MongoClient, req: FindRequest): String {
    val filter = parseFilter(req.filter)
    val projection = parseProjection(req.projection)
    val sort = parseSort(req.sort)
    var cursor = client.getDatabase(req.db).getCollection(req.collection).find(filter)
    if (!projection.isEmpty) cursor = cursor.projection(projection)
    if (!sort.isEmpty) cursor = cursor.sort(sort)
    if (req.skip > 0) cursor = cursor.skip(req.skip)
    if (req.limit > 0) cursor = cursor.limit(req.limit)
    return cursor.explain(ExplainVerbosity.QUERY_PLANNER).toJson(EJSON)
}
